### DISCUSSION TITLE
Private member hooks cannot get  arg on handler functions

### DISCUSSION THREAD

**Discussion by DreamMaoMao** [hyprwm/Hyprland#4168] [General]
when I tried to hook a private function, and the handler function couldn't get the address of the args correctly
![image](https://github.com/hyprwm/Hyprland/assets/30348075/1dd1d246-faec-4de4-b073-b6bb0b5f2b27)
![image](https://github.com/hyprwm/Hyprland/assets/30348075/67d77668-bf7e-4025-8801-aa5b50b6c3b8)

```cpp
static void hkMoveActiveToWorkspace(void* thisptr, std::string args) {
  hygs_log(LOG,"MoveActiveToWorkspace hook toggle:{}",args);
}

void registerGlobalEventHook() {

  static const auto MoveActiveToWorkspaceMethods = HyprlandAPI::findFunctionsByName(PHANDLE, "moveActiveToWorkspace");
  g_pMoveActiveToWorkspaceHook = HyprlandAPI::createFunctionHook(PHANDLE, MoveActiveToWorkspaceMethods[0].address, (void*)&hkMoveActiveToWorkspace);
  g_pMoveActiveToWorkspaceHook->hook();
}
```

**Comment by DreamMaoMao** [comment]
when I don't use the arg,it will work normaly.

![image](https://github.com/hyprwm/Hyprland/assets/30348075/417a88c6-0a60-4238-8508-b52df453f49d)
![image](https://github.com/hyprwm/Hyprland/assets/30348075/8dace323-185f-4e2e-91b3-18323df188b6)

how can i get the arg in handle function of private member hook?

**Comment by vaxerski** [comment]
it's a static member, there is no `thisptr`

**Comment by DreamMaoMao** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/30348075/304a88ae-d960-411d-ab08-cd0eda065d20)

![image](https://github.com/hyprwm/Hyprland/assets/30348075/c696264e-2325-4cd2-bf21-06a15b3106fa)

ok,i have fix it, thanks.

**Comment by DreamMaoMao** [comment]
If thisptr can not be used, then to execute the original function content can only rewrite this function, so it is difficult to keep these function contents in the future hyprland update

Can you consider removing the static of these functions so that the plugin can easily add logic before and after the execution of the function without rewriting the function

@vaxerski

**Reply by vaxerski** [reply]
uhh what?

```
void hkMoveToWorkspace(std::string args) {
     // code before
     original();
     // code after
}
```

**Comment by DreamMaoMao** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/30348075/549b3aa5-6711-42e1-bf28-a101908e6741)
it work, thanks

**Comment by DreamMaoMao** [comment]
I have a strange bug and I need some help. When I try to operate another function  in the same way, it crashes and I can't find the cause of the problem
![image](https://github.com/hyprwm/Hyprland/assets/30348075/0ef9c98b-1d61-4683-a445-7aec2978f8ae)
![image](https://github.com/hyprwm/Hyprland/assets/30348075/3e3d1bf6-95bc-4777-9628-149560b8a94d)

coredump:
![image](https://github.com/hyprwm/Hyprland/assets/30348075/6abb5ac6-ba90-4954-b436-57dc79c1f6f6)

**Comment by DreamMaoMao** [comment]
After using the latest source code, not only can the `fullscreenActive` source function not be executed, but even the hook cannot be mounted.

**Reply by vaxerski** [reply]
> but even the hook cannot be mounted.

you mean .hook() returns false?

most likely this:
![image](https://github.com/hyprwm/Hyprland/assets/43317083/088f1665-63af-4b30-bac8-7b103c66d493)

can you post what the %rip use is? most likely something like `lea [..],(%rip + xxxx)`, that use of %rip is still to be implemented

**Comment by DreamMaoMao** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/30348075/f75bfde7-236e-430b-95ae-73b9c90ddefc)

![image](https://github.com/hyprwm/Hyprland/assets/30348075/b0b9a707-0d0e-4fe9-b12c-64eb09591573)

**Reply by vaxerski** [reply]
oh god, now we got `mov` to implement too? ugh.

**Comment by DreamMaoMao** [comment]
but the fullscreenActive can be hooked in hyprland version "v0.30.0",

The original function of this function cannot be executed in the hook function, this phenomenon exists in the latest source code and v0.30.0 version

**Reply by vaxerski** [reply]
code changes, you know. It depends on how the compiler compiles it.

**Comment by vaxerski** [comment]
@jbeich if I wanted to move to `as` (GNU assembler) and `objcopy` for assembling, how's the situation look like on BSD/Clang?

Basically, I'd use udis like I do right now to generate some asm, e.g.
```asm
mov +0x2cc8c0(%rip), %rdi
```
fix it up
```asm
movabs $0xdeadbeefdeadbeef, %rdx
```
off to a tmp file idk `/tmp/code.asm`
run `as /tmp/code.asm -o /tmp/asm.o && objcopy -O binary -j .text /tmp/asm.o`
read the output binary and remove `/tmp/code.asm` and `/tmp/asm.o`

**Reply by jbeich** [reply]
- Clang has GNU-style assembler integrated into the compiler, so use CC/CXX
- Clang object files may contain LLVM bitcode (when using LTO), so use `llvm-objcopy`
- Clang doesn't generate LLVM bitcode with plain assembly files as opposed to inline assembly in C/C++ files, so non-LLVM `objcopy` may work fine
- FreeBSD doesn't ship `/usr/bin/as` as part of base system but one can install `/usr/local/bin/as` via `binutils` package
- FreeBSD ships both `/usr/bin/objcopy` (ElfToolChain) and `/usr/bin/llvm-objcopy` as part of base system but one can also install `/usr/local/bin/objcopy` via `binutils` package and `/usr/local/bin/llvm-objcopy17` from `llvm17` package (or similar)

**Reply by vaxerski** [reply]
mhm, I getchu. Am kinda busy for the next 2 days (flying back to poland for xmass) but after that I'll make a MR and tag you for bsd/clang verif. Thanks :)

**Comment by DreamMaoMao** [comment]
Is there any progress on this matter?

**Reply by vaxerski** [reply]
no, forgot to do this

**Reply by vaxerski** [reply]
#4256

**Comment by DreamMaoMao** [comment]
it work normally, thanks for you job.
